app/vmalert/web.go

```diff
@@ -87,7 +87,14 @@ func (rh *requestHandler) handler(w http.ResponseWriter, r *http.Request) bool {
 		WriteRuleDetails(w, r, rule)
 		return true
 	case "/vmalert/groups":
-		WriteListGroups(w, r, rh.groups())
+		var data []apiGroup
+		ruleType := r.URL.Query().Get("type")
+		if ruleType == "alert" || ruleType == "record" {
```

Comment:
I don't think we need to have different function for null `ruleType` and non-null `rulesType` in routes, we can pass the `ruleType` to func `rh.groups()` to filter rules, like
```
func (rh *requestHandler) groups(ruleType string) []apiGroup {
	rh.m.groupsMu.RLock()
	defer rh.m.groupsMu.RUnlock()

	groups := make([]apiGroup, 0)
	for _, g := range rh.m.groups {
		g = g.DeepCopy()
		var matchedRules []rule.Rule
		if ruleType != "" {
			for _, r := range g.Rules {
				if _, ok := r.(*rule.AlertingRule); ok && ruleType == "alert" {
					matchedRules = append(matchedRules, r)
				}
				if _, ok := r.(*rule.RecordingRule); ok && ruleType == "record" {
					matchedRules = append(matchedRules, r)
				}
			}
			if len(matchedRules) == 0 {
				continue
			}
			g.Rules = matchedRules
		}
		groups = append(groups, groupToAPI(g))
	}
```
and `rh.listGroups()` will call `rh.groups()` as well, so no need to add below new function.
@victoramsantos wdyt?